src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +276,589 @@ create_distributed_table(PG_FUNCTION_ARGS)
+
+	SplitMode splitMode = NON_BLOCKING_SPLIT;
+	SplitOperation splitOperation = CREATE_DISTRIBUTED_TABLE;
+	SplitShard(
+		splitMode,
+		splitOperation,
+		shardToSplit->shardId,
+		shardSplitPointsList,
+		workersForPlacementList,
+		distributionColumnOverrides,
+		sourceColocatedShardIntervalList,
+		colocationId
+		);
+}
+
+
+/*
+ * SendAddLocalTableToMetadataCommandOutsideTransaction executes metadata add local
+ * table command locally to avoid deadlock.
+ */
+static void
+SendAddLocalTableToMetadataCommandOutsideTransaction(Oid relationId)
+{
+	char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+
+	/*
+	 * we need to allow nested distributed execution, because we start a new distributed
+	 * execution inside the pushed-down UDF citus_add_local_table_to_metadata. Normally
+	 * citus does not allow that because it cannot guarantee correctness.
+	 */
+	StringInfo allowNestedDistributionCommand = makeStringInfo();
+	appendStringInfo(allowNestedDistributionCommand,
+					 "SET LOCAL citus.allow_nested_distributed_execution to ON");
+
+	StringInfo addLocalTableToMetadataCommand = makeStringInfo();
+	appendStringInfo(addLocalTableToMetadataCommand,
+					 "SELECT pg_catalog.citus_add_local_table_to_metadata(%s)",
+					 quote_literal_cstr(qualifiedRelationName));
+
+	List *commands = list_make2(allowNestedDistributionCommand->data,
+								addLocalTableToMetadataCommand->data);
+	char *username = NULL;
+	SendCommandListToWorkerOutsideTransaction(LocalHostName, PostPortNumber, username,
+											  commands);
+}
+
+
+/*
+ * WarnIfTableHaveNoReplicaIdentity notices user if the given table or its partitions (if any)
+ * do not have a replica identity which is required for logical replication to replicate
+ * UPDATE and DELETE commands during create_distributed_table_concurrently.
+ */
+void
+WarnIfTableHaveNoReplicaIdentity(Oid relationId)
+{
+	bool foundRelationWithNoReplicaIdentity = false;
+
+	/*
+	 * Check for source relation's partitions if any. We do not need to check for the source relation
+	 * because we can replicate partitioned table even if it does not have replica identity.
+	 * Source table will have no data if it has partitions.
+	 */
+	if (PartitionedTable(relationId))
+	{
+		List *partitionList = PartitionList(relationId);
+		ListCell *partitionCell = NULL;
+
+		foreach(partitionCell, partitionList)
+		{
+			Oid partitionTableId = lfirst_oid(partitionCell);
+
+			if (!RelationCanPublishAllModifications(partitionTableId))
+			{
+				foundRelationWithNoReplicaIdentity = true;
+				break;
+			}
+		}
+	}
+	/* check for source relation if it is not partitioned */
+	else
+	{
+		if (!RelationCanPublishAllModifications(relationId))
+		{
+			foundRelationWithNoReplicaIdentity = true;
+		}
+	}
+
+	if (foundRelationWithNoReplicaIdentity)
+	{
+		char *relationName = get_rel_name(relationId);
+
+		ereport(NOTICE, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("the relation %s doesn't have a REPLICA "
```

Comment:
super minor:
the relation -> relation (shorter)
doesn't -> does not (more formal)